src/System.Memory/src/System/Buffers/ReadOnlySequence.cs

```diff
@@ -317,29 +343,71 @@ private ReadOnlySequence<T> SliceImpl(SequencePosition begin, SequencePosition e
 
             return new ReadOnlySequence<T>(
                 begin.Segment,
-                begin.Index & IndexBitMask | (Start.Index & ~IndexBitMask),
+                begin.Index & ReadOnlySequence.IndexBitMask | (Start.Index & ReadOnlySequence.FlagBitMask),
                 end.Segment,
-                end.Index & IndexBitMask | (End.Index & ~IndexBitMask)
+                end.Index & ReadOnlySequence.IndexBitMask | (End.Index & ReadOnlySequence.FlagBitMask)
             );
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private SequenceType GetSequenceType()
+        private void GetTypeAndIndices(int start, int end, out SequenceType sequenceType, out int startIndex, out int endIndex)
         {
+            startIndex = start & ReadOnlySequence.IndexBitMask;
+            endIndex = end & ReadOnlySequence.IndexBitMask;
             // We take high order bits of two indexes index and move them
             // to a first and second position to convert to BufferType
             // Masking with 2 is required to only keep the second bit of Start.Index
-            return (SequenceType)((((uint)Start.Index >> 30) & 2) | (uint)End.Index >> 31);
-        }
+            var sequence = (SequenceType)((((uint)Start.Index >> 30) & 2) | (uint)End.Index >> 31);
+            if ((uint)sequence > (uint)SequenceType.String)
+            {
+                ThrowHelper.ThrowInvalidOperationException_UnexpectedSegmentType();
+            }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static int GetIndex(int index) => index & IndexBitMask;
+            sequenceType = sequence;
+        }
 
-        private enum SequenceType
+        private string DebuggerDisplay
         {
-            MemoryList = 0x00,
-            Array = 0x1,
-            OwnedMemory = 0x2
+            // Output:
+            // IMemoryList<Byte>[90] 10 .. 100
+            // OwnedMemory<Byte>[100] 0 .. 100
+            // Char[8] 10 .. 18
+            get
```

Comment:
![image](https://user-images.githubusercontent.com/1142958/36652684-15190f84-1aa8-11e8-8e27-b4e1635d4131.png)

@KrzysztofCwalina seem ok? Had to add it to work out wth was going wrong

Reply:
I'm not a fan, we have a bug to get this fixed, I wouldn't like to slow this PR down by debuggerdisplay discussions.

Reply:
Removed